File: phase-1-dsa-foundations/module-2-core-data-structures/heaps/HeapOperations.java
package heaps;

import java.util.*;

/**
 * Demonstrating Min-Heap and Max-Heap using PriorityQueue.
 */
public class HeapOperations {

    public static void main(String[] args) {
        demonstrateMinHeap();
        demonstrateMaxHeap();
        demonstrateTopKPattern();
    }

    private static void demonstrateMinHeap() {
        System.out.println("--- Min Heap (Default) ---");
        PriorityQueue<Integer> pq = new PriorityQueue<>();

        // Add Elements - O(log n)
        pq.offer(5);
        pq.offer(1);
        pq.offer(10);

        // Peek Min - O(1)
        System.out.println("Min: " + pq.peek()); // 1

        // Poll Min - O(log n)
        while (!pq.isEmpty()) {
            System.out.print(pq.poll() + " "); // 1 5 10
        }
        System.out.println();
    }

    private static void demonstrateMaxHeap() {
        System.out.println("\n--- Max Heap (Custom Comparator) ---");
        // Use Collections.reverseOrder() or lambda (a,b) -> b-a
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());

        pq.offer(5);
        pq.offer(1);
        pq.offer(10);

        System.out.println("Max: " + pq.peek()); // 10

        while (!pq.isEmpty()) {
            System.out.print(pq.poll() + " "); // 10 5 1
        }
        System.out.println();
    }

    private static void demonstrateTopKPattern() {
        System.out.println("\n--- Top K Largest Pattern ---");
        // Find top 3 largest elements
        int[] nums = { 3, 2, 1, 5, 6, 4 };
        int k = 3;

        // Keep Min Heap of size K
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll(); // Remove smallest of the heap
            }
        }
        // Heap contains [4, 5, 6] (Top 3)
        System.out.println("Top " + k + " elements: " + minHeap);
    }
}

File: go/phase-1-dsa-foundations/module-2-core-data-structures/heaps/heap_operations.go
package main

import (
	"container/heap"
	"fmt"
)

// An IntHeap is a min-heap of ints.
type IntHeap []int

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x any) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// MaxHeap wraps IntHeap but inverts Less
type MaxHeap struct{ IntHeap }

func (h MaxHeap) Less(i, j int) bool { return h.IntHeap[i] > h.IntHeap[j] }

func DemonstrateMinHeap() {
	fmt.Println("--- Min Heap ---")
	h := &IntHeap{5, 1, 10}
	heap.Init(h)
	
	fmt.Printf("Min: %d\n", (*h)[0]) // Peek
	
	fmt.Print("Popping: ")
	for h.Len() > 0 {
		fmt.Printf("%d ", heap.Pop(h))
	}
	fmt.Println()
}

func DemonstrateTopK() {
	fmt.Println("\n--- Top K Largest Pattern ---")
	nums := []int{3, 2, 1, 5, 6, 4}
	k := 3
	
	// Keep Min Heap of size K
	h := &IntHeap{}
	heap.Init(h)
	
	for _, num := range nums {
		heap.Push(h, num)
		if h.Len() > k {
			heap.Pop(h) // Remove smallest
		}
	}
	
	// Heap contains top K largest elements
	fmt.Printf("Top %d elements: %v\n", k, *h)
}

func main() {
	DemonstrateMinHeap()
	DemonstrateTopK()
}
